cekit/cli.py

```diff
@@ -1,295 +1,347 @@
+    """
+    run()
+
+
+@build.command(name="osbs", short_help="Build using OSBS engine")
+@click.option('--release', help="Execute a release build", is_flag=True)
+# TODO: Ensure this still makes sense
+@click.option('--tech-preview', help="Execute a tech preview build", is_flag=True)
+@click.option('--user', metavar="USER", help="User used to kick the build as")
+@click.option('--nowait', help="Do not wait for the task to finish", is_flag=True)
+@click.option('--stage', help="Use stage environment", is_flag=True)
+@click.option('--target', metavar="TARGET", help="Override the default target")
+@click.option('--commit-message', metavar="MESSAGE", help="Custom dist-git commit message")
+def build_osbs(release, tech_preview, user, nowait, stage, target, commit_message):
+    """
+    DESCRIPTION
+
+        Executes container image build using OSBS builder.
+
+        https://osbs.readthedocs.io
+    """
+    run()
+
+
+@cli.command(short_help="Execute container image tests")
+@click.option('--steps-url', help="Behave steps library [default: https://github.com/cekit/behave-test-steps.git]", default='https://github.com/cekit/behave-test-steps.git')
+@click.option('--wip', help="Run test scenarios tagged with @wip only", is_flag=True)
+@click.option('--name', 'names', help="Run test scenario with the specified name, can be used specified times", multiple=True)
+@click.argument('image')
+def test(steps_url, wip, names, image):
+    """
+    DESCRIPTION
+
+        Execute container image tests locally using Docker
+
+        NOTE: Image to test must be available in the Docker daemon. It won't be pulled automatically!
+
+    EXAMPLES
+
+        Execute tests for the example/app:1.0 image
+
+            $ cekit test example/app:1.0
+
+        Execute tests that are currently developed and marked with @wip
+
+            $ cekit test --wip example/app:1.0
+
+        Execute specific scenario
+
+            $ cekit test --name 'Check that product labels are correctly set' example/app:1.0
+    """
+
+    if wip and names:
+        raise click.UsageError("Parameters --name and --wip cannot be used together")
+
+    run()
+
+
+def _cli_context_hierarchy(ctx, hierarchy=None):
+    if hierarchy is None:
+        hierarchy = []
+
+    if ctx.parent:
+        _cli_context_hierarchy(ctx.parent, hierarchy)
+
+    hierarchy.append(ctx)
+
+    return hierarchy
+
+
+def run():
+    contexts = _cli_context_hierarchy(click.get_current_context())
+    commands = []
+    params = {}
+
+    for context in contexts:
+        commands.append(context.command.name)
+        params.update(context.params)
 
-    def error(self, message):
-        self.print_help()
-        sys.stderr.write('\nError: %s\n' % message)
-        sys.exit(2)
+    # Remove the default 'cli' command
+    # Conditional here to make tests work better
+    if 'cli' in commands:
+        commands.remove('cli')
 
+    # return Map({'commands': commands, 'params': Map(params)})
 
```

Comment:
Question - why does the cli command need to be removed? ( for the tests ) ?

Reply:
It does not need to be removed in fact, because we test for existence specific commands in the list and `cli` is not one of them.

The comment about the added conditional is that when executed as part of testusite the "default" `cli` command is not added there. I'm not exactly sure now what was the reason. Making it conditional was the easiest way to move forward.

Besides this, I'm still unsure if such parsing will be the final one. I'll give a thought once I get into shape the other CLI (`cekit-cache`).